gprofiler/proc_events.py

```diff
@@ -0,0 +1,159 @@
+    # } exit;
+    _exit_proc_event = struct.Struct("=4I")
+
+    def __init__(self):
+        self._socket = socket.socket(socket.AF_NETLINK, socket.SOCK_DGRAM, self._NETLINK_CONNECTOR)
+        self._exit_callbacks = []
+        self._should_stop = False
+
+        super().__init__(target=self._proc_events_listener, name="Process Events Listener", daemon=True)
+
+    def _register_to_connector_events(self, socket):
+        """Notify the kernel that we're listening to events on the connector"""
+        cn_proc_op = struct.Struct("=I").pack(self._PROC_CN_MCAST_LISTEN)
+        cn_msg = self._cn_msg.pack(self._CN_IDX_PROC, self._CN_VAL_PROC, 0, 0, len(cn_proc_op), 0) + cn_proc_op
+        nl_msg = self._nlmsghdr.pack(self._nlmsghdr.size + len(cn_msg), self._NLMSG_DONE, 0, 0, os.getpid()) + cn_msg
+
+        socket.send(nl_msg)
+
+    def _proc_events_listener(self):
+        """Runs forever and calls registered callbacks on process events"""
+        try:
+            self._socket.bind((os.getpid(), self._CN_IDX_PROC))
+        except socket.error:
+            logger.exception("")
+
+        self._register_to_connector_events(self._socket)
+        selector = selectors.DefaultSelector()
+        selector.register(self._socket, selectors.EVENT_READ)
+
+        while not self._should_stop:
+            events = selector.select()
+            if self._should_stop:
+                break
+
+            for key, _ in events:
+                data = key.fileobj.recv(256)
+
+                nl_hdr = dict(
+                    zip(("len", "type", "flags", "seq", "pid"), self._nlmsghdr.unpack(data[: self._nlmsghdr.size]))
+                )
+                if nl_hdr["type"] != self._NLMSG_DONE:
+                    # Handle only netlink messages
+                    continue
+
+                # Strip off headers
+                data = data[self._nlmsghdr.size : nl_hdr["len"]]
+                data = data[self._cn_msg.size :]
+
+                event = dict(
+                    zip(
+                        ("what", "cpu", "timestamp_ns"),
+                        self._base_proc_event.unpack(data[: self._base_proc_event.size]),
+                    )
+                )
+
+                if event["what"] == self._PROC_EVENT_EXIT:
+                    # (exit_signal is the signal that the parent process received on exit)
+                    event_data = dict(
+                        zip(
+                            ("pid", "tgid", "exit_code", "exit_signal"),
+                            self._exit_proc_event.unpack(
+                                data[
+                                    self._base_proc_event.size : self._base_proc_event.size + self._exit_proc_event.size
+                                ]
+                            ),
+                        )
+                    )
+
+                    for callback in self._exit_callbacks:
+                        callback(event_data["pid"], event_data["exit_code"])
```

Comment:
If there's `pid` and `tgid` - then `tgid` is the "user-space" PID and `pid` is the TID.

In any case - I suggest the API should pass them all? We also want the `exit_signal` because for Java we want to check the `exit_signal`

Reply:
This `pid` is user-space pid. As we discussed yesterday, I'll also parse here the signal out of the exit code.